pkgs/build-support/extract-darwin-app/default.nix

```diff
@@ -0,0 +1,85 @@
+{
+  lib,
+  stdenv,
+  makeWrapper,
+  undmg,
+}:
+
+/** A function to make a derivation from a darwin binary
+  *
+  * It makes the assumption that the binary comes in a ".app" or ".dmg".
+  *
+  * Information about the bundles can be found:
+  * - https://en.wikipedia.org/wiki/Bundle_(macOS)
+  * - https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFBundles/BundleTypes/BundleTypes.html#//apple_ref/doc/uid/10000123i-CH101-SW19
+  * - https://en.wikipedia.org/wiki/Apple_Disk_Image
+  */
+{
+  /** ${appName}.app */
+  appName,
+  /** Name of the file found in ${appName}.app/Contents/MacOS/${binaryName}
+    * to be executed
+    * TODO: detect this automatically from the .plist file?
+    */
+  binaryName,
+  pname,
+  version,
+  /** a nix path or derivation to be used as mkDerivation.src */
+  src,
+  /** is src a .dmg .
+    * These contain multiple folders after extraction
+    * which requires special treatment in the unpackPhase
+    * TODO: Detect this automatically someday
+    */
+  isDmg ? false,
+  /** Whether to use makeWrapper for the binary.
+    * For some packages symbolic links just don't work
+    */
+  wrapBinary ? false,
+  /** The meta attribute passed to mkDerivation */
+  packageMeta ? {},
+  /** Can use a custom function e.g stdenvNoCC.mkDerivation */
+  derivationFunction ? stdenv.mkDerivation,
+}:
+derivationFunction (
+    # .dmg extracts have multiple folders
+    # the unpack phase requires only one folder
+    (lib.attrsets.optionalAttrs isDmg {
+      sourceRoot = "${appName}.app";
+    }) // {
+    inherit pname version src;
+    nativeBuildInputs = (lib.optional wrapBinary makeWrapper) ++ (lib.optional isDmg undmg) ;
+
+    installPhase = ''
+    appDir="$out/Applications/${appName}.app"
+    binDir="$appDir"/Contents/MacOS
+
+    mkdir -p $out/bin
+    if [[ "${toString isDmg}" = "1" ]] ; then
+      # .dmg uses sourceRoot, which means PWD is already in the sourceRoot
```

Comment:
Can't we avoid setting the `sourceRoot` to inside the `.app` bundle so that we can do away with these branches and just copy the bundle in both cases?

Reply:
[2f85c6e](https://github.com/NixOS/nixpkgs/pull/244152/commits/2f85c6eed236dd0c1e523414630a6f6c481e46c7) treats all the usecases above but with the existing branches. I couldn't conceive a solution without branches.

Reply:
Just setting `sourceRoot = "."` should work. It only makes passing an app as the source slightly harder but I don't think that'd be a common case?

Reply:
It does seem like a valid usecase. Is it a problem that it's handled?

Reply:
It makes everything more complex for very little gain though. When would you want to use a straight app as source and be unable to use the directory containing it instead?

Reply:
@michaelCTS, do you disagree strongly with the simplification?